In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
import warnings
from scipy.stats import linregress, uniform
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold, RandomizedSearchCV
from sklearn.compose import make_column_transformer, make_column_selector
from sklearn.impute import KNNImputer, SimpleImputer
from sklearn.preprocessing import FunctionTransformer, OneHotEncoder, StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import f1_score, recall_score, roc_auc_score, precision_score, precision_recall_curve, PrecisionRecallDisplay, confusion_matrix

In [2]:
pd.set_option('display.max_columns', 50)
pd.set_option('display.max_rows', 50)

## 2. Load and Inspect Data

### 2.1. Import and Overview

In [3]:
df = pd.read_csv('data/lead_scoring.csv')
df.head()

Prospect ID  Lead Number              Lead Origin  \
0  7927b2df-8bba-4d29-b9a2-b6e0beafe620       660737                      API   
1  2a272436-5132-4136-86fa-dcc88c88f482       660728                      API   
2  8cc8c611-a219-4f35-ad23-fdfd2656bd8a       660727  Landing Page Submission   
3  0cc2df48-7cf4-4e39-9de9-19797f9b38cc       660719  Landing Page Submission   
4  3256f628-e534-4826-9d63-4a8b88782852       660681  Landing Page Submission   

      Lead Source Do Not Email Do Not Call  Converted  TotalVisits  \
0      Olark Chat           No          No          0          0.0   
1  Organic Search           No          No          0          5.0   
2  Direct Traffic           No          No          1          2.0   
3  Direct Traffic           No          No          0          1.0   
4          Google           No          No          1          2.0   

   Total Time Spent on Website  Page Views Per Visit            Last Activity  \
0                            0                   0.0  Page Visited on Website   
1                          674                   2.5             Email Opened   
2                         1532                   2.0             Email Opened   
3                          305                   1.0              Unreachable   
4                         1428                   1.0        Converted to Lead   

  Country           Specialization How did you hear about X Education  \
0     NaN                   Select                             Select   
1   India                   Select                             Select   
2   India  Business Administration                             Select   
3   India    Media and Advertising                      Word Of Mouth   
4   India                   Select                              Other   

  What is your current occupation  \
0                      Unemployed   
1                      Unemployed   
2                         Student   
3                      Unemployed   
4                      Unemployed   

  What matters most to you in choosing a course Search Magazine  \
0                       Better Career Prospects     No       No   
1                       Better Career Prospects     No       No   
2                       Better Career Prospects     No       No   
3                       Better Career Prospects     No       No   
4                       Better Career Prospects     No       No   

  Newspaper Article X Education Forums Newspaper Digital Advertisement  \
0                No                 No        No                    No   
1                No                 No        No                    No   
2                No                 No        No                    No   
3                No                 No        No                    No   
4                No                 No        No                    No   

  Through Recommendations Receive More Updates About Our Courses  \
0                      No                                     No   
1                      No                                     No   
2                      No                                     No   
3                      No                                     No   
4                      No                                     No   

                                  Tags      Lead Quality  \
0          Interested in other courses  Low in Relevance   
1                              Ringing               NaN   
2  Will revert after reading the email          Might be   
3                              Ringing          Not Sure   
4  Will revert after reading the email          Might be   

  Update me on Supply Chain Content Get updates on DM Content    Lead Profile  \
0                                No                        No          Select   
1                                No                        No          Select   
2                                No                        No  Potential Lead   
3                            

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 37 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Prospect ID                                    9240 non-null   object 
 1   Lead Number                                    9240 non-null   int64  
 2   Lead Origin                                    9240 non-null   object 
 3   Lead Source                                    9204 non-null   object 
 4   Do Not Email                                   9240 non-null   object 
 5   Do Not Call                                    9240 non-null   object 
 6   Converted                                      9240 non-null   int64  
 7   TotalVisits                                    9103 non-null   float64
 8   Total Time Spent on Website                    9240 non-null   int64  
 9   Page Views Per Visit                           9103 

In [5]:
# Describing data
df.describe()

Lead Number    Converted  TotalVisits  Total Time Spent on Website  \
count    9240.000000  9240.000000  9103.000000                  9240.000000   
mean   617188.435606     0.385390     3.445238                   487.698268   
std     23405.995698     0.486714     4.854853                   548.021466   
min    579533.000000     0.000000     0.000000                     0.000000   
25%    596484.500000     0.000000     1.000000                    12.000000   
50%    615479.000000     0.000000     3.000000                   248.000000   
75%    637387.250000     1.000000     5.000000                   936.000000   
max    660737.000000     1.000000   251.000000                  2272.000000   

       Page Views Per Visit  Asymmetrique Activity Score  \
count           9103.000000                  5022.000000   
mean               2.362820                    14.306252   
std                2.161418                     1.386694   
min                0.000000                     7.000000   
25%                1.000000                    14.000000   
50%                2.000000                    14.000000   
75%                3.000000                    15.000000   
max               55.000000                    18.000000   

       Asymmetrique Profile Score  
count                 5022.000000  
mean                    16.344883  
std                      1.811395  
min                     11.000000  
25%                     15.000000  
50%                     16.000000  
75%                     18.000000  
max                     20.000000

### 2.2. Handling `null` values

##### 2.2.1. Handling `Select` value
=> Replace with **Nan**

In [6]:
df = df.replace('Select', np.nan)

#### 2.2.2. Checking null value

In [7]:
#Column wise null values in train data set 
null_perc = pd.DataFrame(round((df.isnull().sum())*100/df.shape[0],2)).reset_index()
null_perc.columns = ['Column Name', 'Null Values Percentage']
null_value = pd.DataFrame(df.isnull().sum()).reset_index()
null_value.columns = ['Column Name', 'Null Values']
null_lead = pd.merge(null_value, null_perc, on='Column Name')
null_lead.sort_values("Null Values", ascending = False)

Column Name  Null Values  \
13             How did you hear about X Education         7250   
28                                   Lead Profile         6855   
25                                   Lead Quality         4767   
31                     Asymmetrique Profile Index         4218   
30                    Asymmetrique Activity Index         4218   
32                    Asymmetrique Activity Score         4218   
33                     Asymmetrique Profile Score         4218   
29                                           City         3669   
12                                 Specialization         3380   
24                                           Tags         3353   
15  What matters most to you in choosing a course         2709   
14                What is your current occupation         2690   
11                                        Country         2461   
9                            Page Views Per Visit          137   
7                                     TotalVisits          137   
10                                  Last Activity          103   
3                                     Lead Source           36   
4                                    Do Not Email            0   
2                                     Lead Origin            0   
1                                     Lead Number            0   
0                                     Prospect ID            0   
20                                      Newspaper            0   
19                             X Education Forums            0   
18                              Newspaper Article            0   
17                                       Magazine            0   
16                                         Search            0   
8                     Total Time Spent on Website            0   
5                                     Do Not Call            0   
6                                       Converted            0   
27                      Get updates on DM Content            0   
26              Update me on Supply Chain Content            0   
22                        Through Recommendations            0   
23         Receive More Updates About Our Courses            0   
21                          Digital Advertisement            0   
34       I agree to pay the amount through cheque            0   
35         A free copy of Mastering The Interview            0   
36                          Last Notable Activity            0   

    Null Values Percentage  
13                   78.46  
28                   74.19  
25                   51.59  
31                   45.65  
30                   45.65  
32                   45.65  
33                   45.65  
29                   39.71  
12                   36.58  
24                   36.29  
15                   29.32  
14                   29.11  
11                   26.63  
9                     1.48  
7                     1.48  
10                    1.11  
3                     0.39  
4                     0.00  
2                     0.00  
1                     0.00  
0                     0.00  
20                    0.00  
19                    0.00  
18                    0.00  
17                    0.00  
16                    0.00  
8                     0.00  
5                     0.00  
6                     0.00  
27                    0.00  
26                    0.00  
22                    0.00  
23                    0.00  
21                    0.00  
34                    0.00  
35                    0.00  
36                    0.00

**Note**: should drop columns with majority of null values:
 Columns with >45% null values:
- `How did you hear about X Education` (78.46%)
- `Lead Profile` (74.19%)
- `Lead Quality` (51.59%)
- `Asymmetrique Profile Index` (45.65%)
- `Asymmetrique Activity Index` (45.65%)
- `Asymmetrique Activity Score` (45.65%)
- `Asymmetrique Profile Score` (45.65%)


### 2.3. Duplicate Check

In [8]:
print("Total number of duplicate values in Prospect ID column :" , df.duplicated(subset = 'Prospect ID').sum())
print("Total number of duplicate values in Lead Number column :" , df.duplicated(subset = 'Lead Number').sum())

Total number of duplicate values in Prospect ID column : 0
Total number of duplicate values in Lead Number column : 0


**Note**: `Prospect ID` and `Lead Number` are unique identicator => drop

## 3. Data Cleaning


#### 3.1. Remove Unnecessary Columns
Drop the columns that are identified above.

`Lead Quality`, `Tags`, `Asymmetrique`  etc. are *created by the sales team once they contact the potential lead* => not be available for the model building.

`Last Notable Activity` is an intermediate column which is an update while the sales team representative is in touch with the lead.

In [9]:
cols_to_drop = ['Prospect ID','Lead Number','How did you hear about X Education','Lead Profile',
                'Lead Quality','Asymmetrique Profile Score','Asymmetrique Activity Score',
               'Asymmetrique Activity Index','Asymmetrique Profile Index','Tags','Last Notable Activity']

In [11]:
df.drop(columns=cols_to_drop, inplace=True)
len(df.columns)

26

### Split train test dataset

In [7]:
train, test = train_test_split(df, test_size=.2, random_state=12, stratify=df['Converted'])
print(f'train shape: {train.shape}')
print(f'test shape: {test.shape}')

train shape: (7392, 37)
test shape: (1848, 37)
